lib/csn.js

```diff
@@ -87,183 +91,144 @@ const getProjectionTarget = entity => isProjection(entity)
+                throw new Error(`Expecting target to be resolved: ${JSON.stringify(elem, null, 2)}`)
+            }
+
+            if (!this.#isPartOfAnyService(draftNodeName)) {
+                LOG.warn(`Ignoring draft node for composition target ${draftNodeName} because it is not part of a service`)
+                continue
+            }
+
+            if (draftNode !== rootEntity && this.#isDraftEnabled(draftNode)) {
+                this.#compileError = true
+                LOG.error(`Composition in draft-enabled entity can't lead to another entity with "@odata.draft.enabled" (in entity: "${entityName}"/element: ${elem.name})!`)
+                delete draftEntities[draftNodeName]
+                continue
+            }
+
+            if (!this.#isDraftEnabled(draftNode) && !draftEntities[draftNodeName]) {
+                this.#collectDraftNodesInto(draftNode, draftNodeName, rootEntity, draftEntities)
+            }
         }
     }
 
     /**
-     * If an entity E is draftable and contains any composition of entities,
-     * then those entities also become draftable. Recursively.
-     * @param {EntityCSN} entity - entity to propagate all compositions from.
+     * @param {EntityCSN} entity - entity to check
+     * @returns {boolean}
      */
-    #propagateCompositions(entity) {
-        if (!this.#getDraftable(entity)) return
-
-        for (const comp of Object.values(entity.compositions ?? {})) {
-            const target = this.#getDefinition(comp.target)
-            const current = this.#getDraftable(target)
-            if (!current) {
-                this.#setDraftable(target, true)
-                this.#propagateCompositions(target)
-            }
-        }
+    #isDraftEnabled(entity) {
+        return entity[DRAFT_ENABLED_ANNO] === true
     }
 
     /** @param {CSN} csn - the full csn  */
-    unroll(csn) {
-        this.csn = csn
+    run(csn) {
+        if (!csn) return
 
-        // inheritance
-        for (const entity of this.#entities) {
-            this.#propagateInheritance(entity)
-        }
+        this.#csn = csn
+        this.#serviceNames = this.#getServiceNames()
+        this.#draftRoots = this.#collectDraftRoots()
 
-        // transitivity through compositions
-        // we have to do this in a second pass, as we only now know which entities are draft-enables themselves
-        for (const entity of this.#entities) {
-            this.#propagateCompositions(entity)
-        }
+        for (const draftRoot of this.#draftRoots) {
+            /** @type {Record<string,EntityCSN>} */
+            const draftEntities = {}
+            this.#collectDraftNodesInto(draftRoot, draftRoot.name, draftRoot, draftEntities)
 
-        this.#propagateProjections()
+            for (const draftNode of Object.values(draftEntities)) {
+                draftEnabledEntities.push(draftNode.name)
+            }
+        }
+        if (this.#compileError) throw new Error('Compilation of model failed')
```

Comment:
why has this become necessary?

Reply:
I wouldn't say necessary. I see it more as an early fail-safe. If one of the following commands with a _invalid_ draft model would be executed 
- `cds compile srv -4 odata`
- `cds compile srv -4 sql`
- `cds w`

you would get the same outcome. The compilation fails.

But if you think typer should only output a log error and just continue (without generating the `drafts` property), then I change the code.

Reply:
I think it is fair to bail in that case. Could you please just add a comment to briefly outline when this could fail? (I.e. like the commands you listed above)